internal/provider/validator_avoid_jaas.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the Apache License, Version 2.0, see LICENCE file for details.
+
+package provider
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+
+	"github.com/juju/terraform-provider-juju/internal/juju"
+)
+
+var _ datasource.ConfigValidator = &AvoidJAASValidator{}
+var _ resource.ConfigValidator = &AvoidJAASValidator{}
+
+// AvoidJAASValidator enforces that the resource is not used with JAAS.
+// Useful to direct users to more capable resources.
+type AvoidJAASValidator struct {
+	Client          *juju.Client
+	PreferredObject string
+}
+
+// Description returns a plain text description of the validator's behavior, suitable for a practitioner to understand its impact.
+func (v AvoidJAASValidator) Description(ctx context.Context) string {
+	return v.MarkdownDescription(ctx)
+}
+
+// MarkdownDescription returns a markdown formatted description of the validator's behavior, suitable for a practitioner to understand its impact.
+func (v AvoidJAASValidator) MarkdownDescription(_ context.Context) string {
+	return "Enforces that this resource should not be used with JAAS"
+}
+
+// ValidateResource performs the validation on the data source.
+func (v AvoidJAASValidator) ValidateDataSource(ctx context.Context, req datasource.ValidateConfigRequest, resp *datasource.ValidateConfigResponse) {
+	resp.Diagnostics = v.validate()
+}
+
+// ValidateResource performs the validation on the resource.
+func (v AvoidJAASValidator) ValidateResource(ctx context.Context, req resource.ValidateConfigRequest, resp *resource.ValidateConfigResponse) {
+	resp.Diagnostics = v.validate()
+}
+
+// validate runs the main validation logic of the validator, reading configuration data out of `config` and returning with diagnostics.
+func (v AvoidJAASValidator) validate() diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	// Return without error if a nil client is detected.
+	// This is possible since validation is called at various points throughout resource creation.
+	if v.Client != nil && v.Client.IsJAAS() {
+		hint := ""
+		if v.PreferredObject != "" {
+			hint = "Try the " + v.PreferredObject + " resource instead."
+		}
+		diags.AddError("Invalid use of resource with JAAS.",
+			"This resource is not supported with JAAS. "+
+				hint+
+				"JAAS offers additional enterprise features through the use of dedicated resources. "+
```

Comment:
The sentence about JAAS is jarring because it shifts the topic from resources (the previous two sentences) to JAAS, and the connection to resources is unclear --  i.e., as a  user, what I'm getting is that dedicated resources can help me take advantage of the additional enterprise features of JAAS, but it's not clear to me what "dedicated resources" is referring to and what I'm supposed to do with this information (i.e., am I suppose to _not_ use the resource suggested in the hint?)

Reply:
I see what you mean. What if we changed this sentence to,
>"JAAS offers additional enterprise features through the use of dedicated resources (those with "jaas" in the name)."

The hint is created above and supposed to lead the user to the correct resource to use - it will say something like "Try the juju_jaas_access_model resource instead."